feat(help): highlighted codeblocks

clason opened the pull request:
Closes #20912 

* bump vimdoc parser and update queries with injections (may need further updating)
* change the syntax file to also recognize codeblock markers with language annotation (necessary)
* add annotations to static part of Nvim Reference Manual (see below)
* adapt `gen_help_html` to updated parser (with bonus syntax highlighting via `highlight.js`, needs https://github.com/neovim/neovim.github.io/pull/303)
* adapt `gen_vimdoc.py` to support language annotations (manually for `lua` and `vim` only)
* add annotations for Lua and C API source code (see below)
* add `news` entry

TODO: 

* backport `syntax/help.vim` change to `release-0.8`

<details><summary>source files with code examples</summary>

* [x] src/nvim/api/vim.c
* [x] src/nvim/api/win_config.c
* [x] src/nvim/api/extmark.c
* [x] src/nvim/api/command.c
* [x] src/nvim/api/autocmd.c
* [x] src/nvim/api/buffer.c
* [x] runtime/lua/vim/diagnostic.lua
* [x] runtime/lua/vim/ui.lua
* [x] runtime/lua/vim/lsp.lua
* [x] runtime/lua/vim/filetype.lua
* [x] runtime/lua/vim/fs.lua
* [x] runtime/lua/vim/keymap.lua
* [x] runtime/lua/vim/_editor.lua
* [x] runtime/lua/vim/shared.lua
* [x] runtime/lua/vim/lsp/handlers.lua
* [x] runtime/lua/vim/lsp/rpc.lua
* [x] runtime/lua/vim/lsp/buf.lua
* [x] runtime/lua/vim/lsp/codelens.lua
* [x] runtime/lua/vim/lsp/util.lua
* [x] runtime/lua/vim/lsp/diagnostic.lua
* [x] runtime/lua/vim/treesitter/query.lua
* [x] runtime/lua/vim/treesitter.lua

</details>

<details><summary>"Nvim reference manual" files</summary>

* [x] develop.txt
* [x] lsp-extension.txt
* [x] if_pyth.txt
* [x] channel.txt
* [x] job_control.txt
* [x] deprecated.txt
* [x] provider.txt
* [x] ui.txt
* [x] dev_style.txt
* [x] treesitter.txt
* [x] api.txt
* [x] support.txt
* [x] diagnostic.txt
* [x] lua.txt
* [x] luaref.txt
* [x] luvref.txt
* [x] news.txt
* [x] nvim.txt
* [x] remote_plugin.txt
* [x] nvim_terminal_emulator.txt
* [x] intro.txt
* [x] lsp.txt
* [x] term.txt
* [x] vim_diff.txt

</details>

clason committed:
feat(help): highlighted codeblocks

clason commented:
OK, this is looking reasonably good on my end so people interested can test this. @justinmk I'd appreciate it if you could give the HTML doc generation a good tire-kicking.

lewis6991 commented:
Looks good to me, however the language name is not concealed:

![image](https://user-images.githubusercontent.com/7904185/203335756-2af29765-d1ba-4d72-a5f6-2edf8ca10395.png)

lewis6991 commented:
With treesitter (as it has language injection).

clason commented:
Doesn't look like an injection to me ;)

lewis6991 commented:
![image](https://user-images.githubusercontent.com/7904185/203336876-54c44492-1e3f-4eb5-b949-19ce1463bee1.png)

clason commented:
Do you have custom conceal?

lewis6991 commented:
Not to my knowledge. You mean a custom setting for `conceallevel`?

clason commented:
Anything, really. I don't see how the `>` can be concealed while the language isn't.

zeertzjq commented:
Won't this look ugly for user who don't have nvim-treesitter installed?

lewis6991 commented:
![image](https://user-images.githubusercontent.com/7904185/203337796-d1ddf2e2-53d5-487e-852c-fc785238c83b.png)

Being concealed. There must be something funky somewhere.

clason commented:
> Won't this look ugly for user who don't have nvim-treesitter installed?

No, because I changed the syntax file to hide the language markers, too (see my second screenshot, which is with standard legacy highlighting).

lewis6991 commented:
Figured it out. I didn't have `syntax` enabled. It was using `conceal` from treesitter.

clason commented:
Still, the queries in this PR should conceal the language as well -- which means you're not using the highlight queries from this PR but either the ones bundled on master or from nvim-treesitter.

lewis6991 commented:
Yep, I think it's pulling in `nvim-treesitter`.

clason commented:
Can you update nvim-treesitter and try again?

lewis6991 commented:
All working now. Looks great!

lewis6991 commented:
For future extra credit, I think applying a different background for codeblocks would make it extra pretty.

lewis6991 commented:
Can't we just apply it to `codeblock>line`, and just make sure the injected highlights have blend?

lewis6991 commented:
With `:hi link @text.literl Folded` it sort of works:

![image](https://user-images.githubusercontent.com/7904185/203342057-3b173380-e2f9-489d-9861-592c2e7f57ab.png)

clason commented:
You'd probably want `@text.literal.help`, but that might be useful. In any case, this is something for (Neovim's future spiffy new default) colorscheme.

clason committed:
docs: add language annotation to Nvim manual

clason committed:
feat(web): syntax highlighting via highlight.js

download from https://highlightjs.org/download/
place `highlight/` directory next to `css/`

style needs adapting for Neovim colors

clason committed:
docs(gen): support language annotation in docstrings

clason commented:
Ok, this is basically done except for the TODOs listed as "to be discussed" @justinmk

craigmac commented:
Off-topic but could we format each codeblock with Stylua or otherwise standardize the examples somehow? Maybe reuse the style config that Neovim core uses? I find that the examples tend to be all over the place and inconsistent use of APIs. For example, some examples have call parentheses and others don't. I think we should also replace usages of `nvim.api.X` in examples when a simpler/more convenient alternative comes out.

clason commented:
This is indeed (very) off-topic here. PRs that improve our documentation are always welcome.

bfredl reviewed (approved):
nicely done!

There are a few instances of incorrect highlights (like vim-blocks that need to be changed from bare eval expressions to full vim script statements), but we can address these in follow up PR:s, as well fine-tuning default highlight etc etc.

bfredl commented on runtime/doc/lua.txt:
```diff
@@ -360,16 +360,16 @@ treated specially.
 From Vimscript the special `v:lua` prefix can be used to call Lua functions
-which are global or accessible from global tables. The expression >
+which are global or accessible from global tables. The expression >vim
     v:lua.func(arg1, arg2)
```
this needs to be prefixed with `call`

bfredl commented on runtime/doc/lua.txt:
```diff
@@ -360,16 +360,16 @@ treated specially.
     return func(...)
-where the args are converted to Lua values. The expression >
+where the args are converted to Lua values. The expression >vim
     v:lua.somemod.func(args)
```
same

bfredl commented on runtime/doc/lua.txt:
```diff
@@ -360,16 +360,16 @@ treated specially.
 
-In addition, functions of packages can be accessed like >
+In addition, functions of packages can be accessed like >vim
     v:lua.require'mypack'.func(arg1, arg2)
     v:lua.require'mypack.submod'.func(arg1, arg2)
```
samesis

bfredl commented on runtime/doc/lua.txt:
```diff
@@ -378,11 +378,11 @@ is still valid as a function call of itself, in case require returns a useful
 The `v:lua` prefix may be used to call Lua functions as |method|s. For
-example: >
+example: >vim
     arg1->v:lua.somemod.func(arg2)
```
and this one

seandewar replied:
(`:eval` for this one)

clason committed:
docs(news): add entry for highlighted code

lewis6991 commented:
anything left blocking this being merged?

clason commented:
> anything left blocking this being merged?

Feedback from @justinmk, and (possibly) merging the linked PR for neovim.io that adds `highlight.js` support for user docs.
